some notes on transaction and block security (#35)
bitmark-inc/docs: bitmark-references/security.markdown

# Security feature of the Bitmark Blockchain

> 

Reviewer: This is another article where we really need a good intro that says what the article is about.

Something like:

"The Bitmark Property System uses signatures and hashes to secure its transactions and blocks."

## Transactions

> 

Reviewer: For the sections, we really need to say how each section relates to the overall topic (of security).

For this section, we might replace the first three lines with something like:

Individual transactions are secured by cryptographic signatures. Single signatures are used for basic transactions, while countersigned signatures are used for special transfers.

There are two types of transactions:
* Single signature - for most basic transactions
* Countersigned - for special transfers

The data in a transaction is packed in a specific order, depending on
the transaction type.  An Ed25519 signature is performed over all data
in the packed transaction and appended to form the submission to the
block chain.

In the case of a countersigned transaction, the signature is performed
over the packed transaction *including* the first signature to form
the complete transaction.  While this does make the signatures order
sensitive, it does mean that the receiver of the transfer can verify
the sender before signing.

The Ed25519 signature algorithm details can be viewed at: https://ed25519.cr.yp.to/

The important features for the Bitmark Block chain are:
* high security signatures with collision resistance
* small 32 byte keys
* signature is only 64 bytes
* fast verification

## Transaction Identifiers

> 

Reviewer: Again, we want to stick with the topic of security.

Something like: "Transactions are securely incorporated into a block through the SHA3-256 algorithm, which is used to derive a 32-byte identifier for each transaction." ...

The SHA3-256 algorithm is used to derive a 32 byte identifier for each
transaction.  This secure hash is over all bytes, including signatures
of the packed transaction.  This is a NIST standard and is the
replacement for SHA2.  SHA3 is used here because it has protection
against the length extension attack that can be performed against
SHA2.

For more details: https://en.wikipedia.org/wiki/SHA-3

The reason for signing over the signature bytes is to protect against
signature substitution.

The Transaction Identifiers are merged into a single root hash using
the Merkle Tree Algorithm: https://brilliant.org/wiki/merkle-tree/

Since the tree is sensitive to missing or changed hashes and to the
order of the hashes and produces a single result it can be used to
minimise the amount of data that is stored in the block header to

> secure the transactions.  The Bitmark Blockchain used SHA3-256 for the

Reviewer 2: Why does `sensitive to the order of hashes` lead to `generate a single result to save disk space`?

Merkle Tree generation.

## Block Identifiers

> 

Reviewer: And once more, we need to push on security to tie everything together. 

Something like:

The SHA3 algorithm also creates security for the blocks themselves, though the Block Identifiers. ...

The Block Identifiers are used both to link the previous block and as
the Proof of Work for the block chain.  Since SHA3 is a fast algorithm
that can be performed on minimal hardware it was decided to use a
memory hard algorithm for the Block Identifiers to prevent the use of
ASIC miners and to set the memory parameters to make GPUs uneconomic
leaving CPUs as the best way to perform the Proof of Work.  This
prevents existing Bitcoin and Litecoin mining hardware from using

> their large system again the chain.

Reviewer 3: s/again/against/g

So the Argon2 was chosen as a modern high security memory hard hashing
algorithm from: https://github.com/P-H-C/phc-winner-argon2

The Block Header is made from the Argon2 Hash of the previous block,
the root Merkle hash covering all transactions and the number of the
block.  Other fields cover timestamp, difficulty and NONCE.  The
difficulty is scaled differently from Bitcoin and allows for much
lower hashing rates; while a modern CPU can run SHA2/SHA3 at many

> thousands of hashes per second, it con only perform one or two Argon2

Reviewer 3: s/con/can/g

> thousands of hashes per second, it con only perform one or two Argon2

Reviewer 3: s/it con/it can/g

hashes per second because of the increase complexity of the Algorithm.

> The NONCE is used for the hashing program to adjust to find a has that

Reviewer 2: ```suggestion
The NONCE is used for the hashing program to adjust to find <what?> that
```
